models/organization.go

```diff
@@ -0,0 +1,218 @@
+	// the corresponding temporal user account that manages this org
+	UserOwner string `gorm:"type:varchar(255);unique"`
+	// the usd value owed by the organization
+	AccountBalance float64 `gorm:"type:float"`
+	// the user accounts who have signed up under this organization
+	RegisteredUsers pq.StringArray `gorm:"type:text[];column:registered_users"`
+}
+
+// OrgManager is an organization model manager
+type OrgManager struct {
+	DB *gorm.DB
+}
+
+// NewOrgManager instantiates an OrgManager
+func NewOrgManager(db *gorm.DB) *OrgManager {
+	return &OrgManager{DB: db}
+}
+
+// NewOrganization is used to create a new organization
+func (om *OrgManager) NewOrganization(name, owner string) error {
+	org := &Organization{
+		Name:      name,
+		UserOwner: owner,
+	}
+	return om.DB.Create(org).Error
+}
+
+// RegisterOrgUser registers an organization user
+func (om *OrgManager) RegisterOrgUser(
+	name,
+	username,
+	password,
+	email string,
+) (*User, error) {
+	// create the user account
+	user, err := NewUserManager(om.DB).NewUserAccount(
+		username,
+		password,
+		email,
+	)
+	if err != nil {
+		return nil, err
+	}
+	// update user model associated organization
+	user.Organization = name
+	// save updated user model
+	if err := om.DB.Model(user).Update(
+		"organization", user.Organization,
+	).Error; err != nil {
+		return nil, err
+	}
+	// update their tier to white-labeled
+	// which will enable organizational based billing
+	if err := NewUsageManager(om.DB).UpdateTier(
+		username,
+		WhiteLabeled,
+	); err != nil {
+		return nil, err
+	}
+	// find organization model
+	org, err := om.FindByName(name)
+	if err != nil {
+		return nil, err
+	}
+	// update organization registered users
+	org.RegisteredUsers = append(org.RegisteredUsers, username)
+	// save updated org model model
+	if err := om.DB.Model(org).Update(
+		"registered_users",
+		org.RegisteredUsers,
+	).Error; err != nil {
+		return nil, err
+	}
+	return user, nil
+}
+
+// FindByName finds an organization by name
+func (om *OrgManager) FindByName(name string) (*Organization, error) {
+	org := &Organization{}
+	if err := om.DB.Where(
+		"name = ?",
+		name,
+	).First(&org).Error; err != nil {
+		return nil, err
+	}
+	return org, nil
+}
+
+// GetOrgUsers is used toreturn the registered users an organization has
+func (om *OrgManager) GetOrgUsers(name string) ([]string, error) {
+	org, err := om.FindByName(name)
+	if err != nil {
+		return nil, err
+	}
+	return org.RegisteredUsers, nil
+}
+
+// IncreaseAccountBalance increases the amount owed by this account
+func (om *OrgManager) IncreaseAccountBalance(name string, amount float64) error {
+	org, err := om.FindByName(name)
+	if err != nil {
+		return err
+	}
+	if org.AccountBalance+amount < org.AccountBalance {
+		return errors.New("account balance overflow error")
+	}
+	org.AccountBalance = org.AccountBalance + amount
+	return om.DB.Model(org).Update("account_balance", org.AccountBalance).Error
```

Comment:
can this be put into a transaction?

Reply:
what do you mean? this kind of stuff is handled by the underlying ORM (in this case, GORM)

Reply:
yup, I put this and decrease account balance into a transaction